docs/proposals/component-registration.md

```diff
@@ -0,0 +1,294 @@
+
+Since it's seemingly impossible to fix Endpoints Problem \#1 without breaking reverse compatibility, this proposal punts on \#1 (until v2 allows reverse-incompatibility) and focuses on fixing \#2 for now. #13778 handles increasing the scope of Endpoints to cover both Ready and NotReady addresses, which is better than nothing, but not as flexible as having an array of conditions, like other -Status API resources.
+
+To fix Endpoints Problem \#2, this proposal moves the specification (and status) of external service provider addresses to its own top-level namespaced API resource (provisionally called **ExternalServiceTarget**). Endpoints Controller will then continue to manage the creation/updates/deletion of Endpoints for internal services based on a selection of Pod instances, but will also do the same for external services based on a selection of ExternalServiceTargets. Endpoints then becomes a (conceptually) read-only API for the user.
+
+
+### API Changes
+
+1. Add new namespaced top level resource: ExternalServiceTarget
+  - Listable, like most other top-level resources
+  - Have both Spec and Status (filterable on both/either)
+  - Spec would have address (ip or host) and ports (optionally named)
+  - Spec would have (optional) LivenessProbe and ReadinessProbe for determining the condition
+  - Status would have an array of [Conditions](../devel/api-conventions.md#typical-status-properties)
+  - Similar to Endpoints, but more static. Defines the set of expected service instances, rather than the set of Ready service instances.
+  - Have to be manually added for non-component external Service instances (e.g. Oracle DB) that do not know about k8s (rather than adding Endpoints)
+  - Can be named with a name generator (similar to pods created by replication controllers) when instances of the same type don't already have unique IDs (generated name is returned in the create response)
+2. Add a new field ServiceSpec.TargetSelector that selects a subset of ExternalServiceTargets
+  - Analogous to the ServiceSpec.Selector-Pod relationship
+3. Add a new ExternalServiceTarget Controller that handles probing ExternalServiceTargets
+  - Probe results update ExternalServiceTarget.Status.Conditions
+4. Modify Endpoints Controller to manage ExternalServiceTarget Endpoints
+  - Readiness from ExternalServiceTarget.Status.Conditions analogous to PodStatus.Conditions
+5. Modify components to self-register ExternalServiceTargets (one per component instance) instead of an Endpoint
+  - Group ExternalServiceTargets (component instances) with labels (e.g. "type=apiserver", "component=core")
+6. Register component Services (with ServiceSpec.TargetSelector) as part of cluster deployment
+  - Services of new components can also be added at runtime, before any instances exist. Each instance would self-register a ExternalServiceTargets.
+7. Modify Service to be support both internal and external targets in the same Service (Pods + ExternalServiceTargets)
```

Comment:
Seems like if we do this, we could just hoist most of the fields at this level into the service.

Service, clusterIP: "None", referencedName: "oracle-db.foo.bar.com", healthCheck / livenessProbe, with a controller that writes health status for referencedName (and maybe endpoints).  And maybe a controller that can look up A records for the referencedName, and updates the endpoints based on the health check on each A record?

The exterior perspective on this proposal is that the general problem of "i want to hook a service up to an external tool not on cluster" is really a first class problem for Kube.  The dependency services have on pods is extremely light - if we broadened the service spec slightly, we'd be able to make external target first class.  I think my concern with adding a new object is that it makes pods "better" than external targets, when I don't think that assumption is warranted.  I want services to solve every use case you describe above.